ENH: Add unstack()

asmeurer opened the pull request:
unstack() is a new function from the 2023.12 version of the array API, which serves as the inverse of stack(), that is, unstack(x, axis=axis) returns a tuple of arrays along axis that when stacked along axis would recreate x.

The implementation is in pure Python, since it is just a straightforward iteration along an index, but if it is preferable it can be moved to an implementation in C.

I haven't yet added any tests.

<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

asmeurer committed:
ENH: Add unstack()

unstack() is a new function from the 2023.12 version of the array API, which
serves as the inverse of stack(), that is, unstack(x, axis=axis) returns a
tuple of arrays along axis that when stacked along axis would recreate x.

The implementation is in pure Python, since it is just a straightforward
iteration along an index, but if it is preferable it can be moved to an
implementation in C.

I haven't yet added any tests.

mhvk commented on numpy/_core/shape_base.py:
```diff
@@ -455,6 +457,77 @@ def stack(arrays, axis=0, out=None, *, dtype=None, casting="same_kind"):
+    np.True_
+
+    """
+    x = asanyarray(x)
```
I think turning things into an array here is overkill, as it unnecessarily makes the code not work on array ducktypes -- converting lists of lists here seems not particularly reasonable!

mhvk commented on numpy/_core/shape_base.py:
```diff
@@ -455,6 +457,77 @@ def stack(arrays, axis=0, out=None, *, dtype=None, casting="same_kind"):
+    """
+    x = asanyarray(x)
+
+    axis = normalize_axis_index(axis, x.ndim)
```
In principle, one can just do `return tuple(np.rollaxis(x, axis))`.

seberg replied:
Right, this whole function is just `tuple(np.moveaxis(x, axis, 0))`.

There is no reason to have it NumPy beyond torch not supporting the `tuple()` I think.
(@rgommers let's make sure, because if they do it seems unnecessary.  This is a function that if we did customize `ndarray.__array_namespace__`, I may not include in NumPy.

Although it is also one of those functions that can be useful for discoverability (using moveaxis like that is obvious to us, but not for every user)!  In that sense, I do think it may be a nice teaching opportunity to say that it is equivalent!

rgommers replied:
> There is no reason to have it NumPy beyond torch not supporting the `tuple()` I think.

https://github.com/data-apis/array-api/issues/487 covers reasons, as pointed out by @asmeurer.  

> I do think it may be a nice teaching opportunity to say that it is equivalent!

That might be useful in the Notes section for a few people, but I don't think we should be encouraging using `tuple()` on an array at all (nor `len()` for that matter).

seberg replied:
Ah, right, there is always the argument that array iteration isn't necessary well defined (i.e. my `iteraxis()` thoughts, also)!

mhvk commented on numpy/_core/shape_base.py:
```diff
@@ -455,6 +457,77 @@ def stack(arrays, axis=0, out=None, *, dtype=None, casting="same_kind"):
+    dstack : Stack arrays in sequence depth wise (along third axis).
+    column_stack : Stack 1-D arrays as columns into a 2-D array.
+    vsplit : Split an array into multiple sub-arrays vertically (row-wise).
+    unstack : Split an array into a tuple of sub-arrays along an axis.
```
Should not do `see also` to oneself! In general, I'd tend to remove quite a number of the items here: `[vhd]stack`, definitely `column_stack`, and probably `vsplit`.

mhvk reviewed (commented):
Not sure how useful this is, but of course good to stay consistent with array API.

asmeurer commented:
For reference, here was the discussion for adding this to the array API https://github.com/data-apis/array-api/issues/487

asmeurer committed:
Cleanup the unstack see-also list

asmeurer committed:
Simplify the implementation of unstack()

asmeurer committed:
Add a notes section to the unstack docstring

asmeurer committed:
Add a test for unstack()

asmeurer committed:
Add unstack to Sphinx and cleanup the docstring

asmeurer commented:
Addressed comments and added tests and added the docstring to Sphinx. 

As with #26572 I have no idea how to run the doctests. The refguide_check script doesn't seem to actually run the doctests for this function.

ev-br commented:
> I have no idea how to run the doctests.

With https://github.com/numpy/numpy/pull/26604

```
$ spin smoke-docs  -v |grep stack
numpy/__init__.py::numpy.vstack PASSED                                   [  0%]
numpy/__init__.py::numpy.column_stack PASSED                             [  1%]
numpy/__init__.py::numpy.hstack PASSED                                   [  2%]
numpy/__init__.py::numpy.row_stack PASSED                                [  6%]
numpy/__init__.py::numpy.stack PASSED                                    [ 10%]
numpy/__init__.py::numpy.dstack PASSED                                   [ 15%]
^C
```

(spin integration is not complete yet, `-t path/to/file` does not work yet)

mtsokol reviewed (commented):
LGTM, only typing stubs + typing tests and a release note is missing.

asmeurer committed:
Add type overloads and test for unstack

asmeurer committed:
Add release notes entry for unstack

asmeurer commented:
I added typing stubs and a changelog entry. You'll need to take a close look at the typing stuff because I don't really know what I'm doing there, so there's a good chance it's wrong.

mtsokol commented on numpy/_core/shape_base.py:
```diff
@@ -455,6 +457,72 @@ def stack(arrays, axis=0, out=None, *, dtype=None, casting="same_kind"):
     return _nx.concatenate(expanded_arrays, axis=axis, out=out,
                            dtype=dtype, casting=casting)
 
+def _unstack_dispatcher(x, *, axis=None):
```
Should be `_unstack_dispatcher(x, /, *, axis=None)`

mtsokol replied:
Done!

mtsokol commented on numpy/typing/tests/data/reveal/shape_base.pyi:
```diff
@@ -53,3 +53,6 @@ assert_type(np.kron(AR_f8, AR_f8), npt.NDArray[np.floating[Any]])
 assert_type(np.tile(AR_i8, 5), npt.NDArray[np.int64])
 assert_type(np.tile(AR_LIKE_f8, [2, 2]), npt.NDArray[Any])
+
+assert_type(np.unstack(AR_i8, 5), list[npt.NDArray[np.int64]])
+assert_type(np.unstack(AR_LIKE_f8, 5), list[npt.NDArray[Any]])
```
Let's change `5` to `axis=0` - it's a keyword only argument.

mtsokol replied:
Done!

mtsokol commented on numpy/_core/shape_base.pyi:
```diff
@@ -117,6 +117,12 @@ def stack(
     casting: _CastingKind = ...
 ) -> _ArrayType: ...
 
+@overload
+def unstack(
+    array: _ArrayLike[_SCT],
+    axis: int = 0
+) -> tuple[NDArray[_SCT], ...]: ...
+
```
```python
@overload
def unstack(
    array: _ArrayLike[_SCT],
    /,
    *,
    axis: int = ...,
) -> tuple[NDArray[_SCT], ...]: ...
@overload
def unstack(
    array: ArrayLike,
    /,
    *,
    axis: int = ...,
) -> tuple[NDArray[Any], ...]: ...
```

mtsokol replied:
Done!

mtsokol committed:
Review changes

mtsokol committed:
Fix reveal test

mtsokol committed:
Fix reveal test

jakevdp commented on numpy/_core/shape_base.py:
```diff
@@ -455,6 +457,74 @@ def stack(arrays, axis=0, out=None, *, dtype=None, casting="same_kind"):
+    >>> np.all(arr == arr2)
+    np.True_
+
+    """
```
Should we error here if `x.ndim == 0` rather than having the error originate from `moveaxis`?

mtsokol replied:
Right, I think we can error here.

mtsokol replied:
Done!

mtsokol committed:
Raise for 0-d input

mattip commented:
Thanks @asmeurer
